smm/lotus_adapter.go

```diff
@@ -0,0 +1,266 @@
+package smm
+
+import (
+    "context"
+    "github.com/filecoin-project/lotus/api"
+    "github.com/filecoin-project/lotus/build"
+    "github.com/filecoin-project/lotus/chain/actors"
+    "github.com/filecoin-project/lotus/chain/actors/aerrors"
+    laddress "github.com/filecoin-project/lotus/chain/address"
+    "github.com/filecoin-project/lotus/chain/store"
+    "github.com/filecoin-project/lotus/chain/types"
+    "github.com/ipfs/go-cid"
+)
+
+type lotusAdapter struct {
+    address      laddress.Address
+    fullNode     api.FullNode
+    listeners    map[StorageMiningEvents]struct{}
+    headChanges  <-chan[]*store.HeadChange
+    context      context.Context
+}
+
+func NewNode(ctx context.Context, fullNode api.FullNode, address Address) (Node, error) {
+    adapter := lotusAdapter{
+        fullNode: fullNode,
+        listeners: make(map[StorageMiningEvents]struct{}), // this needs synchronization
+        context: ctx,
+    }
+    var err error
+    adapter.address, err = laddress.NewFromString(string(address))
+    if err != nil {
+        return nil, err
+    }
+    adapter.headChanges, _ = fullNode.ChainNotify(ctx)
+    go adapter.eventHandler()
+    return adapter, nil
+}
+
+// TODO: Implement this for TipSetKey
+func stateChangesFromHeadChanges(changes []*store.HeadChange) (stateChanges []*StateChange, err error) {
```

Comment:
I gather you're planning to implement this when the tipset key lands and before merging this PR. 

My understanding from a quick skim of the store code is that this channel will have a value for every tipset in sequence, backwards then forwards. I think the intention for the module interface is to receive only a single new state for reverts, aggregating a sequence (and only if there is one), and then a single new state for the final state. So the return should have at most two values if the input has one contiguous sequence of reverts, and then one sequence of forward applications.